### PR TITLE
Make CallOptions and AbstractStub use Deadline.

### PR DESCRIPTION
**Pull request by buchgr**
- Made CallOptions use the Deadline type instead of a
  long to represent a deadline.
- Added new methods CallOptions.withDeadline(Deadline) and
  AbstractStub.withDeadline(Deadline). The methods are
  marked experimental, as the Deadline class is marked
  experimental. These methods are meant to replace
  CallOptions.withDeadlineNanoTime(Long) and
  AbstractStub.withDeadlineNanoTime(Long), which have
  been deprecated.
- Updated CallOptions.toString() to include all fields.

### CODE REVIEW

**Comment by buchgr** [comment]
PTAL @ejona86 @louiscryan

File: core/src/main/java/io/grpc/CallOptions.java [hunk 1]
```diff
@@ -103,10 +118,12 @@ public CallOptions withCompression(@Nullable String compressorName) {
    * @param deadlineNanoTime the deadline in the clock as per {@link System#nanoTime()}.
    *                         {@code null} for unsetting the deadline.
    */
+  @Deprecated
   public CallOptions withDeadlineNanoTime(@Nullable Long deadlineNanoTime) {
-    CallOptions newOptions = new CallOptions(this);
-    newOptions.deadlineNanoTime = deadlineNanoTime;
-    return newOptions;
+    Deadline deadline = deadlineNanoTime != null
+        ? Deadline.after(deadlineNanoTime - System.nanoTime(), TimeUnit.NANOSECONDS)
```
Comment (buchgr) [inline core/src/main/java/io/grpc/CallOptions.java]: is the potential for inaccuracy here a problem? Same question for `getDeadlineNanoTime()`.

File: core/src/main/java/io/grpc/CallOptions.java [hunk 2]
```diff
@@ -103,10 +118,12 @@ public CallOptions withCompression(@Nullable String compressorName) {
    * @param deadlineNanoTime the deadline in the clock as per {@link System#nanoTime()}.
    *                         {@code null} for unsetting the deadline.
    */
+  @Deprecated
   public CallOptions withDeadlineNanoTime(@Nullable Long deadlineNanoTime) {
```
Comment (buchgr) [inline core/src/main/java/io/grpc/CallOptions.java]: I deprecated the `*NanoTime(...)` methods, as they haven't been marked experimental.

Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/CallOptions.java]: FWIW, they can probably never go away.

Methods marked deprecated need to point to the correct way to do something in their comments.

File: core/src/main/java/io/grpc/CallOptions.java (hunk 1 above)
Comment (ejona86) [inline core/src/main/java/io/grpc/CallOptions.java]: The inaccuracy doesn't bother me. That sort of thing is done every time you call ScheduledExecutor.schedule. This is also deprecated, so I'm not bothered by some minor inaccuracy.

File: core/src/main/java/io/grpc/Deadline.java [hunk 3]
```diff
@@ -139,4 +139,9 @@ public long timeRemaining(TimeUnit unit) {
     Preconditions.checkNotNull(scheduler, "scheduler");
     return scheduler.schedule(task, deadlineNanos - System.nanoTime(), TimeUnit.NANOSECONDS);
   }
+
+  @Override
+  public String toString() {
+    return String.valueOf(deadlineNanos);
```
Comment (ejona86) [inline core/src/main/java/io/grpc/Deadline.java]: Doing "X ns from now" may not be a bad idea. Just the nanos is useless without a reference to the clock.

Comment (buchgr) [inline core/src/main/java/io/grpc/Deadline.java]: `Deadline.toString()` for sure is useful. But what should it look like? The problem I see with this implementation is, if someone logs a `Deadline` object, only deadlines from the same JVM process can be compared meaningfully, as the `System.nanoTime()` part of `deadlineNanos` is kinda random. 

Should I implement this to return a proper date/time instead?

Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/Deadline.java]: It would be nice if the units were printed out too.  We have a class to do this for timeouts in GrpcUtil that is almost what you want.

Comment (ejona86) [inline core/src/main/java/io/grpc/Deadline.java]: I wouldn't mind if it was:
```java
return "" + timeRemaining(NANOSECONDS) + "ns from now";
```

GrpcUtil is another option. I think it would probably be more useful to be relative instead of printing out the date/time, but date/time would be more helpful than just the raw deadlineNanos.

**Comment by ejona86** [comment]
@buchgr LGTM

**Comment by ejona86** [comment]
codecov/patch is non-binding. It can be okay to submit if it fails, but you should at least consider if you should add more tests. Reaching 90-95% coverage requires constant vigilance!

File: core/src/main/java/io/grpc/Deadline.java (hunk 3 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/Deadline.java]: FWIW I just implemented a timer heap in a personal project of mine, and I agree with Eric: relative times are really useful, absolute times are harder to read.  Since log statements typically contain a timestamp anyways, relative doesn't seem so bad.

File: core/src/main/java/io/grpc/CallOptions.java (hunk 2 above)
Comment (lukaszx0) [inline core/src/main/java/io/grpc/CallOptions.java]: +1 to what @carl-mastrangelo said. Let's add `@Depreacted` javadoc and point to `withDeadline` ?

File: core/src/main/java/io/grpc/Deadline.java (hunk 3 above)
Comment (buchgr) [inline core/src/main/java/io/grpc/Deadline.java]: @carl-mastrangelo good point!

**Comment by buchgr** [comment]
@carl-mastrangelo @ejona86 PTAL

File: core/src/main/java/io/grpc/CallOptions.java (hunk 1 above)
Comment (buchgr) [inline core/src/main/java/io/grpc/CallOptions.java]: @ejona86 but `getDeadlineNanoTime()` now returns `System.nanoTime()` if the deadline is expired, but it would return some value less than `System.nanoTime()` before.

Depending on how people use these methods (e.g. to compare values), it might break their code i.e. one would expect the below to always be true
```java
opts.withDeadlineNanoTime(nanos).getDeadlineNanoTime() == nanos
```

We could retain the original behaviour by making the `Deadline` constructor package private and introduce a package private method `Deadline.deadlineNanos()`, which we immediately deprecate as well. WDYT?

Comment (ejona86) [inline core/src/main/java/io/grpc/CallOptions.java]: The other option is to change Deadline.timeRemaining() to return negative numbers. That would not result in nanos being equal, but it would be close. It does seem it could be useful to know how far in the past a deadline expired, which seems like an argument for returning negative numbers from timeRemaining(). The argument against returning negative numbers is applications may be buggy in their handling of it.

File: core/src/test/java/io/grpc/CallOptionsTest.java [hunk 4]
```diff
@@ -108,27 +114,39 @@ public void mutateExecutor() {
...
-    assertEquals("CallOptions{deadlineNanoTime=null, authority=null}",
-        CallOptions.DEFAULT.toString());
+    assertEquals("CallOptions{deadline=null, authority=null, "
+        + "affinity={}, executor=null, compressorName=null}", CallOptions.DEFAULT.toString());
+
     // Deadline makes it hard to check string for equality.
-    assertEquals("CallOptions{deadlineNanoTime=null, authority=authority}",
-        allSet.withDeadlineNanoTime(null).toString());
-    assertTrue(allSet.toString().contains("deadlineNanoTime=" + sampleDeadlineNanoTime + ","));
+    assertEquals("CallOptions{deadline=null, authority=authority, "
```
Comment (carl-mastrangelo) [inline core/src/test/java/io/grpc/CallOptionsTest.java]: Just check that the string .contains("123ns form now")

**Comment by carl-mastrangelo** [comment]
LGTM

File: core/src/main/java/io/grpc/CallOptions.java (hunk 1 above)
Comment (buchgr) [inline core/src/main/java/io/grpc/CallOptions.java]: @ejona86 yep :). any preference? I just wanna make sure this change doesn't break any applications. //cc @carl-mastrangelo

Comment (ejona86) [inline core/src/main/java/io/grpc/CallOptions.java]: @buchgr, returning negative numbers came up in the initial commit of `timeRemaining()`, but @louiscryan chose not to return negatives. I talked with @louiscryan about what you are experiencing, and we're both in favor of having `timeRemaining()` return negatives. So let's go that way. It won't be 100% accurate, but the edge case is racy no matter how you look at it.

Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/CallOptions.java]: +1 (-1?) to negative numbers.

File: core/src/test/java/io/grpc/CallOptionsTest.java (hunk 4 above)
Comment (buchgr) [inline core/src/test/java/io/grpc/CallOptionsTest.java]: I do, 3 lines down :).

Comment (carl-mastrangelo) [inline core/src/test/java/io/grpc/CallOptionsTest.java]: Ah.  In that case this line should be changed to check for absence rather than exact string matching.

File: core/src/test/java/io/grpc/CallOptionsTest.java [hunk 5]
```diff
@@ -108,27 +118,55 @@ public void mutateExecutor() {
...
+        + "affinity={sample=blah}, executor=class io.grpc.internal.SerializingExecutor, "
+        + "compressorName=null}",
+        allSet.withDeadline(null)
+            .withExecutor(new SerializingExecutor(MoreExecutors.directExecutor())).toString());
+
+    long remainingNanos = extractRemainingTime(allSet.toString());
+    long delta = TimeUnit.MILLISECONDS.toNanos(10);
+    assertNotNull(allSet.getDeadline());
+    assertEquals(remainingNanos, allSet.getDeadline().timeRemaining(NANOSECONDS), delta);
```
Comment (buchgr) [inline core/src/test/java/io/grpc/CallOptionsTest.java]: @carl-mastrangelo added a regex to match for `Deadline.toString` as well as filter the nanotime. WDYT?

File: core/src/test/java/io/grpc/CallOptionsTest.java (hunk 4 above)
Comment (buchgr) [inline core/src/test/java/io/grpc/CallOptionsTest.java]: well it does check for absence by `deadline=null`?

**Comment by buchgr** [comment]
@ejona86 @carl-mastrangelo PTAL

Should be good to go now. Modified `Deadline.timeRemaining()` to return a negtive offset in a separate commit. Added some more tests for `toString()` in `CallOptionsTest` and `DeadlineTest` that uses a regex to better check the deadline.

File: core/src/main/java/io/grpc/Deadline.java [hunk 6]
```diff
@@ -114,18 +114,15 @@ public Deadline offset(long offset, TimeUnit units) {
   /**
    * How much time is remaining in the specified time unit. Internal units are maintained as
    * nanoseconds and conversions are subject to the constraints documented for
-   * {@link TimeUnit#convert}. If there is no time remaining {@code 0} is always returned.
+   * {@link TimeUnit#convert}. If there is no time remaining, a negative offset from the
```
Comment (ejona86) [inline core/src/main/java/io/grpc/Deadline.java]: Mentioning "system clock" is a bit strange, although it could be fine. You might consider something like, "If there is no time remaining, the returned duration is how long ago the deadline expired."

File: core/src/test/java/io/grpc/CallOptionsTest.java (hunk 4 above)
Comment (carl-mastrangelo) [inline core/src/test/java/io/grpc/CallOptionsTest.java]: The idea here is to reduce burden when updating CallOptions.  If these tests spuriously fail because the string representation intended solely for debugging was slightly off, that's bad.  Checking for just the snippet you care about in the string not only solves this, but make it more explicit what the test is actually expecting.  It's a maintenance win.

File: core/src/test/java/io/grpc/CallOptionsTest.java [hunk 7]
```diff
@@ -108,27 +118,55 @@ public void mutateExecutor() {
...
+    CallOptions opts = CallOptions.DEFAULT.withDeadlineNanoTime(System.nanoTime());
+    assertNotNull(opts.getDeadlineNanoTime());
+    assertTrue(opts.getDeadlineNanoTime() <= System.nanoTime());
+    assertNotNull(opts.getDeadline());
+    long delta = MILLISECONDS.toNanos(10);
+    assertEquals(0, opts.getDeadline().timeRemaining(NANOSECONDS), delta);
+    assertTrue(opts.getDeadline().isExpired());
+  }
+
+  private static long extractRemainingTime(String deadlineStr) {
```
Comment (carl-mastrangelo) [inline core/src/test/java/io/grpc/CallOptionsTest.java]: This seems like overengineering, but meh.

File: core/src/test/java/io/grpc/CallOptionsTest.java [hunk 8]
```diff
@@ -108,27 +118,55 @@ public void mutateExecutor() {
...
+            .withExecutor(new SerializingExecutor(MoreExecutors.directExecutor())).toString());
+
+    long remainingNanos = extractRemainingTime(allSet.toString());
+    long delta = TimeUnit.MILLISECONDS.toNanos(20);
+    assertNotNull(allSet.getDeadline());
+    assertEquals(remainingNanos, allSet.getDeadline().timeRemaining(NANOSECONDS), delta);
+  }
+
+  @Test
+  @SuppressWarnings("deprecation")
```
Comment (ejona86) [inline core/src/test/java/io/grpc/CallOptionsTest.java]: A better approach may be to just mark the test as `@Deprecated`.

**Comment by ejona86** [comment]
@buchgr LGTM

File: core/src/test/java/io/grpc/CallOptionsTest.java (hunk 7 above)
Comment (buchgr) [inline core/src/test/java/io/grpc/CallOptionsTest.java]: tiny bit :smiley:

**Comment by buchgr** [comment]
thanks guys! addressed the comments. will squash and merge once CI is done.

**Comment by ejona86** [comment]
@buchgr, if Visual Studio 2013 passes, that's commonly "good enough" and not worth waiting for Travis. That doesn't mean ignore Travis if it fails though :smile:

**Comment by buchgr** [comment]
cherry-picked (a4640c88ceb4777c0ba2d83eb3e2861e1da2030c) and (40575bc9fa99a8c69ad23943f287917008cd91b9).

@ejona86 I ll make sure to blame a broken build on you :stuck_out_tongue_winking_eye:
